extensions/amp-pinterest/0.1/amp-pinterest.js

```diff
@@ -0,0 +1,574 @@
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+/**
+ * @fileoverview Shows a Pinterest widget.
+ * Examples:
+ * <code>
+ *
+ * <amp-pinterest height=20 width=40
+ *   data-do="buttonPin"
+ *   data-url="http://www.flickr.com/photos/kentbrew/6851755809/"
+ *   data-media="http://farm8.staticflickr.com/7027/6851755809_df5b2051c9_z.jpg"
+ *   data-description="Next stop: Pinterest">
+ * </amp-pinterest>
+ *
+ * <amp-pinterest width=239 height=400
+ *   data-do="embedBoard"
+ *   data-scale-height="289"
+ *   data-scale-width="107"
+ *   data-url="http://www.pinterest.com/kentbrew/art-i-wish-i-d-made/">
+ * </amp-pinterest>
+ *
+ * </code>
+ */
+
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {loadPromise} from '../../../src/event-helper';
+
+const VALID_PARAMS = [
+    'data-config',
+    'data-height',
+    'data-width',
+    'data-shape',
+    'data-color',
+    'data-lang',
+    'data-scale-width',
+    'data-scale-height',
+    'data-board-width' ];
+
+const POP = 'status=no,resizable=yes,scrollbars=yes,' +
+  'personalbar=no,directories=no,location=no,toolbar=no,' +
+  'menubar=no,width=900,height=500,left=0,top=0';
+
+const BASE60 = '0123456789ABCDEFGHJKLMNPQRSTUVWXYZ_abcdefghijkmnopqrstuvwxyz';
+
+class AmpPinterest extends AMP.BaseElement {
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+  /** @override */
+  layoutCallback() {
+
+    let height = this.element.getAttribute('height');
+
+    let width = this.element.getAttribute('width');
+
+    let pinDo = AMP.assert(this.element.getAttribute('data-do'),
+      'The data-do attribute is required for <amp-pinterest> %s',
+      this.element);
+
+    let pinUrl = AMP.assert(this.element.getAttribute('data-url'),
+      'The data-url attribute is required for <amp-pinterest> %s',
+      this.element);
+
+    // make a 12-digit base-60 number for performance tracking
+    let guid = '';
+    for (let i = 0; i < 12; i = i + 1) {
+      guid = guid + BASE60.substr(Math.floor(Math.random() * 60), 1);
+    }
+
+    // open a new window
+    let pop = function (url, shouldPop) {
+      if (shouldPop) {
+        // amp=1&guid=guid are already in the query before long fields
+        window.open(url, '_pinit', POP);
+      } else {
+        // this is a straight window.open; append amp=1&guid=guid
+        window.open(url + '?amp=1&guid=' + guid, '_blank');
+      }
+    };
+
+    // log a string
+    let log = function (str) {
+      let call = new Image();
+      let query = 'https://log.pinterest.com/?guid=' + guid;
+      query = query + '&amp=1';
+      // add optional string &foo=bar
+      if (str) {
+        query = query + str;
+      }
+      query = query + '&via=' + document.URL;
```

Comment:
encodeURIComponent of the URL?

window.location.href is more idiomatic.

use this.element.ownerDocument.defaultView to get the right window.

We might eventually reuse these across iframes.